### ISSUE TITLE
Expose Typescript type definition

### ISSUE THREAD

**Issue by lutangar** [Chocobozzz/PeerTube#4201]
**Describe the problem to be solved**

In order to develop a __PeerTube__ plugin in Typescript, I should be able to import __PeerTube__ plugin API type definitions.

**Describe the solution you would like:**

Writing a plugin could look like the following :
```ts
import { RegisterServerOptions } from '@peertube/plugin';

export async function register({ registerSetting }: RegisterServerOptions) {
  registerSetting({
    // ...
  });
}

export async function unregister() {
  return;
}
```

Publishing to [https://github.com/DefinitelyTyped/DefinitelyTyped](DefinitelyTyped) is another option:
```ts
import { RegisterServerOptions } from '@types/peertube';
```
Or maybe:
```ts
import { RegisterServerOptions } from '@types/peertube-plugin';
```

I don't know if it makes much sense to publish either part of `client` or `server` as a npm package.
But if it does I would consider publishing plugin typedef with these.
For ex:
- `@peertube/client`
- `@peertube/server`

> Just so you know there is an npm package named `peertube` with no connection with this project:
> https://www.npmjs.com/package/peertube

**Describe alternatives you have considered**  

If it doesn't make sense the alternatives would be to either create another `package.json` in `./server/types/plugins` for example.
Or maybe create a dedicated directory in this repository and publish it to https://github.com/DefinitelyTyped/DefinitelyTyped#create-a-new-package

As for the best name of such package, I'd love to hear other opinions, because I'm still unsure as to weather it would make sense to expose typedef other than the plugin ones.

**Comment by JohnXLivingston** [comment]
My plugin https://github.com/JohnXLivingston/peertube-plugin-livechat is completely written in Typescript (backend and frontend).

I had discussions with @Chocobozzz  on how we can have type files. Our conclusions was something like: I will write a npm package that brings Peertube's type by manually copying them (we found no easy way to export them automatically).
But I did not manage to do that. Especially because I have no clue on how to properly handle types depending on peertube's versions. First idea was to synchronise types version with peertube version. But in my plugin I want to be compatible with multiple versions of peertube...
So I gave up.

What I did instead: I wrote my own type files in my plugin : 
https://github.com/JohnXLivingston/peertube-plugin-livechat/blob/main/client/%40types/peertube.d.ts
https://github.com/JohnXLivingston/peertube-plugin-livechat/blob/main/server/%40types/peertube.d.ts

In the current version, I dropped backward compatibility with Peertube <3.2.0. But in older versions of my plugin, you can see comments for old peertube compatibility. For example: https://github.com/JohnXLivingston/peertube-plugin-livechat/blob/31c10621d1821c7bba78fbaca042afbca70cdf2b/server/%40types/peertube.d.ts#L134
See how some attributes are optional when not available in every Peertube version.

You can freely copy my definitions files if you are interested.

**Comment by lutangar** [comment]
@JohnXLivingston regarding your version issues, I suppose it's related to this point :

> Unfortunately, we don't have enough resources to provide hook compatibility between minor releases of PeerTube (for example between 1.2.x and 1.3.x).
> https://docs.joinpeertube.org/contribute-plugins?id=compatibility-with-peertube

The compatibility warning is about hooks here, but I suppose there might be other compatibility issues on other parts of the plugin API...

*...that being said,  @Chocobozzz I understand maintaining backward compatibility between minors may be a constraint, even if it can't be guaranteed, even if it's never reached, it should be set as a goal nonetheless.*

> See how some attributes are optional when not available in every Peertube version.

@JohnXLivingston given better backward compatibility and maybe a more flexible version constraints on the `engine.peertube` (see https://github.com/npm/node-semver#usage for example)  I think you could mitigate most of your issues.

Now about these type definitions... I described a few possible paths in my first comment, on how it could achieved in this very repository and I still strongly think it should __and__ I'm willing to help make it possible. Ofc the idea is to streamline it with the release process, so maintainers won't have to think about this. 

We mostly need to figure out the name of such package, and choose the right folder organization.

**Comment by lutangar** [comment]
@Chocobozzz I'd love to hear your opinion on this.

> We mostly need to figure out the name of such package, and choose the right folder organization.

What's your feeling about the naming issue?

**Comment by Chocobozzz** [comment]
Something like `@peertube/peertube-typing`?

**Comment by buoyantair** [comment]
It's also possible to directly export typedefinitions as part of the project via something like `index.d.ts`. A lot of projects now export type definitions without needing to publish a separate package at @types. Webpack's own library is itself a good example for this https://github.com/webpack/webpack

**Comment by lutangar** [comment]
Hey there, just noticed npm package [@peertube/peertube-types](https://www.npmjs.com/package/@peertube/peertube-types) wasn't published yet. 

I thought it was published since #4531 was merged. Once published we can close this issue.

**Comment by Chocobozzz** [comment]
Hello,

I just published https://www.npmjs.com/package/@peertube/peertube-types (4.0.0-beta.1). Please test and create issues if there is something wrong with the package.
